### PR TITLE
Bugfix/dap mouse broken#367

### PR DESCRIPTION
**Pull request by nbfalcon**
Make dap's hover tooltips not broken. Please see the REVIEW comments also.

### CODE REVIEW

File: dap-mouse.el [hunk 1]
```diff
@@ -197,32 +197,35 @@ This function must return nil if it doesn't handle EVENT."
...
+                      ;; perhaps do proper error handling?
+                      (= request-id dap-tooltip--request))
+             (add-text-properties
+              start end '(mouse-face dap-mouse-eval-thing-face))
+             ;; REVIEW: I inherited this code; why kill the buffer first,
+             ;; then show it, and then initialize it?
+             (when (get-buffer dap-mouse-buffer)
+               (kill-buffer dap-mouse-buffer))
+             (-let [(&hash "result" "variablesReference" variables-reference) var]
+               (if (= variables-reference 0)
```
Comment (yyoncho) [inline dap-mouse.el:213]: use zerop

File: dap-mouse.el [hunk 2]
```diff
@@ -197,32 +197,35 @@ This function must return nil if it doesn't handle EVENT."
...
+                   (message result)     ; primitive value -> posframe awkward
```
Comment (yyoncho) [inline dap-mouse.el:214]: why is that? If it is primitive it still can be displayed in the posframe.

File: dap-mouse.el (hunk 1 above)
Comment (nbfalcon) [reply dap-mouse.el]: That line is gone, and I used zerop everywhere else.

File: dap-mouse.el (hunk 2 above)
Comment (nbfalcon) [reply dap-mouse.el]: Fixed it.

**Comment by nbfalcon** [comment]
There is one more issue: for some reason, after hovering, the buffer gets marked as modified.

File: dap-mouse.el [hunk 3]
```diff
@@ -197,32 +197,36 @@ This function must return nil if it doesn't handle EVENT."
...
-               (message message))))
+         (-lambda ((&hash "success" "body" (var &as &hash? "result")))
+           (when (and success
+                      ;; REVIEW: hover failure will yield weird errors involving
+                      ;; process filters, so I resorted to this hack; we should
+                      ;; perhaps do proper error handling?
+                      (= request-id dap-tooltip--request))
+             (add-text-properties
+              start end '(mouse-face dap-mouse-eval-thing-face))
+             ;; REVIEW: I inherited this code; why kill the buffer first,
```
Comment (yyoncho) [inline dap-mouse.el:208]: Without this, in some cases, the posframe was not calculating the proper size of the tooltip.

**Review by yyoncho — APPROVED** [review]

File: dap-mouse.el (hunk 3 above)
Comment (nbfalcon) [reply dap-mouse.el:208]: Ahh, that makes sense. Thank you.

**Comment by nbfalcon** [comment]
Dont merge yet.

**Comment by nbfalcon** [comment]
How can I debug json messages between the server and Emacs? For some reason, when hovering over a variable that isn't defined yet, ptvsd sends a success response + empty result. This would cause empty hover boxes, so I check if the result isn't empty. This way, no empty box is shown, but the identifier is still highlighted. Should I change that?

**Comment by nbfalcon** [comment]
@yyoncho, what do you think? Should I keep this workaround, should I perhaps also not highlight the identifier?

**Comment by yyoncho** [comment]
> @yyoncho, what do you think? Should I keep this workaround, should I perhaps also not highlight the identifier?

I dont see this as a big issue. 

> How can I debug json messages between the server and Emacs? 

Set dap-print-io to t

**Comment by nbfalcon** [comment]
@yyoncho I have it set to t, where would I find its result buffer? (Not restarted Emacs). Also, the last issue: why does the debugged buffer get marked as modified when I hover over a variable?

**Comment by kiennq** [comment]
The log should be spewed on `*Message*`

**Comment by nbfalcon** [comment]
Thanks, I can now conclude that I was right: ptsvd is weird. Here is the response for a hover on a yet-to-be-defined variable:
```json
{
  "type": "response",
  "seq": 21,
  "request_seq": 15,
  "success": true,
  "command": "evaluate",
  "message": "",
  "body": {
    "result": "",
    "variablesReference": 0,
    "presentationHint": {}
  }
}
```
"success": true, "variablesReference" = 0 -> no children and "result" is the empty string...

**Comment by nbfalcon** [comment]
All that is missing now is to determine why the debugged buffer is modified when showing the posframe. Any ideas, anyone? Perhaps it is related to doom?

**Comment by nbfalcon** [comment]
Im at a loss for my last point. Perhaps @kieenq might have some more ideas on how to not mark the buffer as modified.

**Comment by kiennq** [comment]
> Im at a loss for my last point. Perhaps @kieenq might have some more ideas on how to not mark the buffer as modified.

There's `(set-buffer-modified-p FLAG)` if you want to mark the buffer as not modified. Can you try with vanilla Emacs to see if it's problem of `doom` or not?

**Comment by nbfalcon** [comment]
I give up; slightly broken mouse support is still better than completely broken mouse support, so I'll merge it anyway.

**Pull request merged**